Foreman provider refresh

kbrock opened the pull request:
Foreman Refresh working from the provider level.

- [x] pull out entities like `Manager` into own pr.
-  ~~pull out save_inventory_assoc into separate pr~~
- [x] enable refresh per manager
- [x] link from database not children and managers (think this is referring to providers)
-  ~~create `managiq-foreman` gem~~ merge in `manageiq-foreman` code


/cc @brandondunne @gmcculloug

Fryguy commented on vmdb/app/models/ems_refresh/manager.rb:
```diff
@@ -0,0 +1,19 @@
+      end
+    end
+
+    module InstanceMethods
```
This isn't necessary anymore, and is generally avoided.  You can just use normal module mixin methods for instance methods.

kbrock replied:
I got deprecation warnings, so avoided is actually saying it lightly. thanks.

Fryguy commented:
cc @blomquisg

Fryguy commented on vmdb/app/models/ems_refresh/parsers/foreman.rb:
```diff
@@ -0,0 +1,88 @@
+      result[:configuration_profiles] = configuration_profile_inv_to_hashes(inv[:hostgroups], ids)
+      add_ids(ids, result[:configuration_profiles])
+      result[:configured_systems] = configured_system_inv_to_hashes(inv[:hosts], ids)
+      result
```
Forgot to mention I was hoping this could use the `@data` and `@data_index` concept from cloud refreshers, but oh well.  That pattern is what we are considering for rewriting parsers into using a builder-like syntax.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/foreman.rb:
```diff
@@ -0,0 +1,88 @@
+    def media_inv_to_hashes(media)
+      media.map do |m|
+        {
+          "manager_ref"  => "medium:#{m["id"]}",
```
The manager_ref should be also close to the metal as possible without any adjustment.  So just `m["id"]` here is sufficient.  Unless this overlaps with the ptables (i.e. the ids could overlap).

kbrock replied:
this overlaps with the ptables.
any other ideas? There is a guid column, but the id would be best for us for single host refresh. Their apis are tuned to the active record id.

Fryguy replied:
this is tricky one.  You *could* make the key on type and manager_ref in the save_inventory.  We've had to do mutli-key checks in other places, and that should work fine.

Otherwise, this is fine.  It makes it kind of weird because there are patterns elsewhere to access native objects by ems_ref, so you would need to do something different and parse the ems_ref before using it.  For example, there is a method with_provider_object that could be defined on CustomizationScriptMedium, and internally it would use ems_ref directly.

kbrock replied:
1. I don't like the current implementation.
2. I did see `name:#{id}` used in another provider
3. It is going to be tricky to support this in general. agree that it is best to not modify fields
4. foreman is not consistent with their naming. so you'll notice `operatingsystem_id` vs `operating_system`, making it even worse.
5. I don't like the multi key ids, but I suppose they may be necessary.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/foreman.rb:
```diff
@@ -0,0 +1,88 @@
+    def add_ids(target, recs, key = "manager_ref")
+      recs.each_with_object(target) do |rec, dest|
+        dest[rec[key]] = rec
+      end
```
Why the need for each_with_object?

```ruby
recs.each { |r| target[r[key]] = r }
```

kbrock replied:
the configuration_profile points (line 74) to the hash of the configuration profile. It uses the ids to put it there.

So after the configuration profile is saved, the id from that hash can be used to populate `configuration_profile_id`.

Fryguy replied:
I'm not questioning the method...I'm questioning the implementation...my example above does the same thing with less code and without calling each_with_object.

Fryguy commented:
Why the v2 qualifier on the spec?  Are there multiple APIs we are using?

Fryguy commented:
Looks pretty good...I know you are breaking this up to do manager level refresh so I'm holding off on commenting on that part of this PR.

kbrock commented:
@brandondunne and I felt it was necessary to distinguish.
But I guess no real reason. Any comments brandon?

brandondunne commented:
@Fryguy We assumed that we will eventually be adding a spec for a v3 api, so we decided to add v2 to the name now.

jvlcek commented:
This PR is not yet complete enough for the build team to comment on.

Fryguy commented on vmdb/app/models/ems_refresh/refreshers/foreman_configuration_refresher.rb:
```diff
@@ -0,0 +1,21 @@
+  module Refreshers
+    class ForemanConfigurationRefresher < BaseRefresher
+      include EmsRefresherMixin
+      include RefresherRelatsMixin
```
Do you need this?

Fryguy commented on vmdb/app/models/ems_refresh/save_inventory_helper.rb:
```diff
@@ -99,4 +99,28 @@ def store_ids_for_new_records(records, hashes, keys)
+  # most of the refresh_multi calls follow the same pattern
+  # this pulls it out
+  def save_inventory_assoc(type, parent, hashes, target, find_key, child_keys = [], extra_keys = [])
+    reflection = parent.reflections[type]
```
Prefer the method reflect_on_association, since it's Rails 3/4 agnostic.  Right @tenderlove ?

brandondunne commented on vmdb/app/models/ems_refresh/parsers/foreman.rb:
```diff
@@ -0,0 +1,119 @@
+
+      private
+
+      def fetch_provisioning_manager_ids(manager)
```
@kbrock I feel like this is saver logic, do you?  Just going based on the name, I would expect the parser to parse the raw data into a structure that the saver will accept.  Then I would expect the saver to deal with database interaction (id's, relationships, etc.).  If that was the case, the parser could live in the gem as you originally wanted, right?  @Fryguy what do you think?

kbrock replied:
@brandondunne Yes, I was able to move out. Moved to the method that calls this. will look at moving into the saver

brandondunne commented on vmdb/app/models/ems_refresh/refreshers/foreman_refresher.rb:
```diff
@@ -0,0 +1,13 @@
+    class ForemanRefresher
+      def self.refresh(providers)
+        providers.each do |provider|
+          EmsRefresh.refresh(provider.provisioning_manager)
```
Why the provisioning manager instead of the configuration manager?

kbrock replied:
```
foreman(provider) -> foreman_provisioning_manager -> foreman_configuration_manager
```

Since most of the time we want to update the `configuration_manager` and not the `provisioning_manager`, refreshing the `configuration_manager` instead may make more sense.
The first case would be a little slower, but the expectations are that it will be a little slower, so that may be fine.

brandondunne commented on vmdb/spec/models/ems_refresh/refreshers/foreman_refresher_spec.rb:
```diff
@@ -0,0 +1,96 @@
+
+  it "will perform a full refresh on api v2" do
+    VCR.use_cassette("#{described_class.name.underscore}_api_v2") do
+      EmsRefresh.refresh(provider)
```
In most cases we run the refresh twice before checking the data to verify that the second run doesn't change the numbers (creating duplicate records).  We'll probably want to do something similar, but maybe run configuration_manager, check some values, run provisioning_manager, check some values, run configuration_manager again and verify that relationships now exist and values didn't change.

brandondunne commented:
@kbrock looking great so far!

Fryguy commented on lib/manageiq_foreman/Gemfile:
```diff
@@ -0,0 +1,6 @@
+# Specify your gem's dependencies in manageiq_foreman.gemspec
+gemspec
+
+# gem "dothtml", path: "../dothtml"
```
? I think you can just remove this.

Fryguy commented on lib/manageiq_foreman/lib/manageiq_foreman/paged_response.rb:
```diff
@@ -0,0 +1,42 @@
+        @size    = @results.size
+        @page    = json["page"]
+      else # Array
+        @results = json.kind_of?(Hash) ? [json] : json
```
How can this ternary ever be true?

kbrock replied:
foreman comes back with 2 possible responses:

```
{ total: 3, results: [ { id: 5}, {id:6},{id:7} ] } # case 1

{ id: 5, name: x} # case 2
```

The third use case is when we have a response and we want to `collect` it into another object, but want to keep the paging information.

```
if case 1
  results = json["results"]
else # case 2 or 3
  results = x ? case 2 : case 3
end
```

kbrock replied:
since we introduced `map!`, maybe we no longer need case 3.
Not sure if it will creep up when provisioning.

Fryguy replied:
Oh the && in the first clause mixed with the nested ternary for the clause was just confusing...especially since the else has that misleading `# Array` comment.

I saw `if json.kind_of?(Hash)` and assumed the else meant "not a Hash", but that extra && kills that.

blomquisg replied:
Could this line just be:

```ruby
    @results = Array.wrap(json)
```

kbrock replied:
@blomquisg I could have sworn `Array[hash]` didn't work, but it does - thanks

Fryguy commented on lib/manageiq_foreman/spec/paged_response_spec.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  describe "#map!" do
+    subject(:paged_response) do
```
Why subject?

kbrock replied:
I guess subject and let do the same thing

typically subject is not named.
I can convert to let if you want

Fryguy replied:
`subject { stuff }`
is equiv to
`let(:subject) { stuff }`

And `subject(:foo) { stuff }`
is equiv to
`let(:subject) { stuff }`
`let(:foo) { subject }`

So if you don't use the actual subject keyword, then there's no need to call it a subject.

Fryguy commented on vmdb/app/models/customization_script.rb:
```diff
@@ -1,4 +1,7 @@
   belongs_to :provisioning_manager
+
+  scope :ptables, where(:type => "CustomizationScriptPtable")
+  scope :media, where(:type => "CustomizationScriptMedium")
```
I think it's preferable to just use normal class methods instead of scope.  i.e. A class method that returns an AR::Relation is a scope.

kbrock replied:
these don't work as methods.

I'll see what else I can do

matthewd replied:
Err, they should work as methods :confused:

Personally I like scopes for this sort of thing, partly just because it provides a natural grouping, and partly because it back-stops you with a guarantee that you'll end up returning an AR::Relation.

But if they're going to be scopes, they should be using lambdas (forward compatible, and generally more predictable behaviour):

```ruby
scope :media, lambda { where(:type => "CustomizationScriptMedium") }
```

Fryguy replied:
The scopes are weird to begin with...why would a caller need to do `CustomizationScript.ptables.where()` when they can just do `CustomizationScriptPtable.where()`

matthewd replied:
@Fryguy it's `some_customization_script.ptables.where(...)`

Fryguy replied:
@matthewd Is `scope` at the class level or instance level?  I thought it was the former.

> But if they're going to be scopes, they should be using lambdas (forward compatible, and generally more predictable behaviour):

Yeah, I think if you don't do it with a lambda you get a deprecation warning in Rails 4.

BTW, needing the lambda is explicitly why I like the methods better.  It's like scope is a useless DSL for something already in Ruby (i.e. taking a block of code and attaching it to a name, a.k.a. defining a method).  Unless you can info on them with something like `SomeModel.scopes`?

kbrock replied:
They are just used in the specs, so I can move the logic over there.

matthewd replied:
Sorry, yeah, I confused myself entirely. They are indeed class methods.

The one feature is that your lambda can return nil, and it'll do the right thing. I wouldn't consider that enough of a reason to create such a concept, but as it's there... to me, it helps distinguish the simple named filters from the possibly-numerous other class methods. Certainly once your scope starts taking parameters, then it's almost always going to be more confusing than helpful.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/foreman.rb:
```diff
@@ -0,0 +1,116 @@
+        recs.collect do |cs|
+          {
+            "type"                       => "ConfiguredSystemForeman",
+            "manager_ref"                => "host:#{cs["id"]}",
```
Do you need the "host" differentiation?  If, not then it's preferable to just use cs["id"] directly.

kbrock replied:
ok, how do you want to handle configuration scripts? (ptable vs medium) ?

Fryguy replied:
That is the exception because we are merging two disparate things from the raw data.  Here it's just one thing.  Even so, what are the ids from ptable vs medium.  Are they potentially identical?  Even so, does it matter?  If you refresh against the ptables association separately from the medium association, then you automatically don't collide, unless I"m missing something.

kbrock replied:
We need to lookup `medium` and `ptable` by `manager_ref`. They are both stored in the `configuration_scripts` table. The ids are rails ids. So yes, they most probably overlap.

I could look them up by type and id. But it just adds a whole bunch of complexity and verbiage.
I can separate the ids when we are pulling it out, I'll have to put the scope back in then :)

Fryguy commented on vmdb/app/models/ems_refresh/parsers/foreman.rb:
```diff
@@ -0,0 +1,116 @@
+        recs.collect do |profile|
+          {
+            "type"                           => "ConfigurationProfileForeman",
+            "manager_ref"                    => "hostgroup:#{profile["id"]}",
```
Same ehre.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/foreman.rb:
```diff
@@ -0,0 +1,116 @@
+        key = record[id_key]
+        return unless key
+        ids["#{prefix}:#{key}"].tap do |v|
+          @missing_key = true unless v
```
So it's one common flag for everything?  Seems wrong, unless I"m missing something.

Fryguy replied:
Oh nevermind...this is more of flag to say that we encountered at least one missing key...I didn't realize that at first.

blomquisg replied:
I wonder if this needs to be something more descriptive.

I originally read this as, "there's some key missing...".  But, it only ever records that *at least* one key is missing.

Later I realized it's used to queue up a refresh of another manager in the same provider.  Maybe it should be something more like that.  Maybe `refresh_other_managers`.  Or, something that says why this boolean is being set ... why is it needed at a higher level.

Fryguy commented on vmdb/app/models/ems_refresh/refreshers/foreman_configuration_refresher.rb:
```diff
@@ -0,0 +1,29 @@
+
+      def save_inventory(manager, targets, hashes)
+        EmsRefresh.save_configuration_manager_inventory(manager, hashes, targets[0])
+        EmsRefresh.refresh(manager.provider.provisioning_manager, targets[0]) if hashes[:missing_key]
```
Should be .queue_refresh, right?

blomquisg commented on lib/manageiq_foreman/lib/manageiq_foreman/inventory.rb:
```diff
@@ -0,0 +1,45 @@
+
+    # generate inventory for the foreman provider
+    def refresh_all(target = nil)
+      refresh_configuration(target).merge(refresh_provisioning)
```
Should target be passed to `refresh_provisioning`, too?

Fryguy replied:
Is target even needed?

Fryguy commented on vmdb/app/models/ems_refresh/refreshers/foreman_provisioning_refresher.rb:
```diff
@@ -0,0 +1,21 @@
+
+      def save_inventory(manager, targets, hashes)
+        EmsRefresh.save_provisioning_manager_inventory(manager, hashes, targets[0])
+        EmsRefresh.refresh(manager.provider.configuration_manager)
```
Same with the `queue_refresh`

Fryguy commented on vmdb/app/models/ems_refresh/save_inventory_configuration.rb:
```diff
@@ -0,0 +1,34 @@
+    end
+
+    def save_configuration_profiles_inventory(manager, hashes, target)
+      hash_contains_all_values = target.nil? || manager == target
```
I don't get the variable name here.  How does targeting the whole manager make it "contains all values"

kbrock replied:
That is the way I read the current logic: If the `target` is the `ems`, then we are loading all relations into the hash.

If the hash contains all children, then lets run through the `deleted = parent.assoc.dup` scenarios.

you like `delete_missing_records` better?

Fryguy replied:
oh i see...yeah, it's confusing because this stuff is introducing the targeting concept, even though targeting doesn't exist for anything but vmware.

yeah, delete_missing_records reads much nicer, or maybe `process_all_deletes` or something.

Fryguy commented on vmdb/app/models/ems_refresh/save_inventory_configuration.rb:
```diff
@@ -0,0 +1,34 @@
+    def save_configured_systems_inventory(manager, hashes, target)
+      hash_contains_all_values = target.nil? || manager == target
+      hashes.each do |hash|
+        hash[:configuration_profile_id] = hash.fetch_path("configuration_profile", :id)
```
I would expect the incoming has to be key based, not string based.

blomquisg commented on lib/manageiq_foreman/lib/highline_helper.rb:
```diff
@@ -0,0 +1,53 @@
+include HighLine::SystemExtensions
+
+# stolen from appliance_console/prompts.rb
+module HighlineHelper
```
What's this for?  I don't see it pulled in anywhere, and I'm not sure how it's used.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/foreman.rb:
```diff
@@ -0,0 +1,116 @@
+      def operating_system_flavors_inv_to_hashes(flavors_inv, ids)
+        flavors_inv.collect do |os|
+          {
+            "manager_ref"           => "operating_system:#{os["id"]}",
```
This too should just be the id...We should only qualify the ids where we have to, like in the customization scripts table where we blend objects.  Even so, I know I asked this before, but are those ids in the customization scripts table likely to be unique even across types, or could they overlap.

Fryguy commented on vmdb/app/models/ems_refresh/parsers/foreman.rb:
```diff
@@ -0,0 +1,116 @@
+                                                                              inv[:customization_scripts])
+        add_ids(ids, result[:configuration_profiles])
+        result[:configured_systems] = configured_system_inv_to_hashes(inv[:hosts], ids, inv[:operating_system_flavors])
+        result[:missing_key] = true if missing_key
```
Won't this key mess up the save_inventory stuff?  I never see it removed anywhere.  I think it's fine here, but should be removed before being passed into save_*_inventory.

Fryguy commented on vmdb/app/models/ems_refresh/save_inventory_helper.rb:
```diff
@@ -99,4 +99,27 @@ def store_ids_for_new_records(records, hashes, keys)
+  # This generates the "deletes" values based upon this intent
+  def relation_values(parent, reflection, target)
+    # always want to refresh this association
+    reflection = parent.class.reflect_on_association(reflection) if reflection.kind_of?(Symbol)
```
Why the Symbol check?  That method accepts string or symbol.

kbrock replied:
This allowed you to pass in a symbol or an association.
Changed to just look it up.

Fryguy commented on vmdb/app/models/ems_refresh/save_inventory_helper.rb:
```diff
@@ -99,4 +99,27 @@ def store_ids_for_new_records(records, hashes, keys)
+  def relation_values(parent, reflection, target)
+    # always want to refresh this association
+    reflection = parent.class.reflect_on_association(reflection) if reflection.kind_of?(Symbol)
+    top_level = reflection.options[:dependent] == :destroy
```
I don't get the dependent => destroy check.  What does that have to do with updating an association?

kbrock replied:
Some associations (provision_manager -> operating_system_flavor) are the definitive source of the association. We want to delete those base records if they are missing. So we declare `:dependent => :destroy`.

Other associations (configuration_profile => :operating_system_flavor) are references, but we don't want to delete the flavor from the db if it is removed. So we don't declare `:dependent => :destroy`.

The same logic that goes for active record goes for the refresh.

I called it a `top_level` association - it owns the lifespan of the child objects.

In our code we tend to just hardcode the logic into the ruby. But I thought we could just leverage AR.

I'm pretty sure this is universal in our refreshes, but we can make sure when/if we actually integrate this.

kbrock replied:
added this to the code per @blomquisg's  comment

Fryguy commented on vmdb/spec/models/ems_refresh/refreshers/foreman_refresher_spec.rb:
```diff
@@ -0,0 +1,113 @@
+      "name"        => "ProviderRefreshSpec-Media",
+      "type"        => "CustomizationScriptMedium",
+      "manager_ref" => "medium:8"
+    )
```
We have a special matcher for this which you might have seen from the other specs...this can be:

```ruby
expect(medium).to have_attributes(
  :name        => "ProviderRefreshSpec-Media",
  :type        => "CustomizationScriptMedium",
  :manager_ref => "medium:8"
)

Fryguy commented on vmdb/spec/models/ems_refresh/refreshers/foreman_refresher_spec.rb:
```diff
@@ -0,0 +1,113 @@
+    expect(osf.customization_scripts).to include(medium)
+  end
+
+  # these values are from the
```
I thikn your comment got truncated.

Fryguy commented:
Since the manageiq_foreman directory has its own spec directory, how are those specs run in the automated tests?

blomquisg commented on vmdb/app/models/ems_refresh/save_inventory_helper.rb:
```diff
@@ -99,4 +99,27 @@ def store_ids_for_new_records(records, hashes, keys)
     end
   end
+
+  # most of the refresh_multi calls follow the same pattern
```
`refresh_multi` should be `save_inventory_multi`

blomquisg commented on vmdb/app/models/ems_refresh/save_inventory_helper.rb:
```diff
@@ -99,4 +99,27 @@ def store_ids_for_new_records(records, hashes, keys)
+  # We need to determine our intent:
+  # - make a complete refresh. Delete missing values.
+  # - make a partial refresh. Don't delete missing keys
+  # This generates the "deletes" values based upon this intent
```
Maybe add a comment here on how intent is derived from these parameters.  Below, it looks like it's based on:

```ruby
   top_level && (target == true || target.nil? || parent == target)
```

But, I wanna see that expressed in English to know how the parent, the reflection, and the target are being used to determine the intent.  Esp. since the target can have so many possibilities.

kbrock commented:
removed hash problems and most of spec rubocops - some complaints just can't be solved.

kbrock commented:
@Fryguy how do you want me to fix lib specs so they are run? Just move into the spec tree and abandon keeping this together like a gem? I'm guessing you and @blomquisg have thoughts how this will be done when providers are moved around. Maybe @matthewd has an idea here?

jvlcek commented:
The build team is OK with this PR. The new gem and it's runtime dependencies will be manageable.

kbrock commented:
@Fryguy tests now running in lib's specs.
Anything else you can find?

Fryguy commented on lib/Gemfile:
```diff
@@ -76,5 +77,9 @@ unless ENV['APPLIANCE']
   end
 end
+
+gem 'pry'
```
This should be removed...it will end up affecting the build team.  Additionally, I don't want pry to become a dependency for ManageIQ.  That is what the Gemfile.dev.rb file is for.

Fryguy commented on lib/manageiq_foreman/lib/manageiq_foreman/connection.rb:
```diff
@@ -0,0 +1,119 @@
+        break if small.empty?
+        all += small.to_a
+      end
+      PagedResponse.new(all)
```
You have a paged_response method down below, but don't use it?  Also why the need for that method...`PagedResponse.new(all)` is just as clear as `paged_response(all)`

Fryguy commented on lib/manageiq_foreman/lib/manageiq_foreman/connection.rb:
```diff
@@ -0,0 +1,119 @@
+module ManageiqForeman
+  class Connection
```
I feel like there needs to be a public/private split somewhere in this class.

Fryguy commented on lib/manageiq_foreman/lib/manageiq_foreman/inventory.rb:
```diff
@@ -0,0 +1,36 @@
+      {
+        :operating_systems => connection.all(:operating_system_details),
+        :media             => connection.all(:media),
+        :ptables           => connection.all(:ptables),
```
Not needed for now, but can these be collected in parallel?  I know when we did the RHEV refresh we got enormous performance gains by parallelizing the collection.

brandondunne commented on lib/Gemfile:
```diff
@@ -40,6 +40,7 @@ gem "uuidtools",            "~>2.1.3",       :require => false
 gem "sass",                 "3.1.20",        :require => false
 gem "trollop",              "~>1.16.2",      :require => false
 gem "psych",                "~>2.0.10"
+gem "foreman_api",          "~>0.1.11",      :require => false  # used by lib/manageiq_foreman
```
Should this be moved to the manageiq-foreman Gemfile?

Fryguy replied:
We can't reference it as a gem (build reasons), so somehow it has to be referenced here.

brandondunne commented on lib/Gemfile:
```diff
@@ -76,5 +77,9 @@ unless ENV['APPLIANCE']
     gem "timecop",    "~>0.5.3",  :require => false
     gem "xml-simple", "=1.0.12",  :require => false  # Used by test/xml/tc_xmlhash_methods.rb
+    gem "vcr",        "~>2.6",    :require => false  # Used by manageiq-foreman
+    gem "webmock",                :require => false  # Used by vcr / manageiq-foreman
```
Same here

Fryguy replied:
I get this warning locally:

```
Your Gemfile lists the gem vcr (~> 2.6) more than once.
You should probably keep only one of them.
While it's not a problem now, it could cause errors if you change the version of just one of them later.
Your Gemfile lists the gem webmock (>= 0) more than once.
You should probably keep only one of them.
While it's not a problem now, it could cause errors if you change the version of just one of them later.
```

If they are moving here, can you remove them from vmdb Gemfile as well?

kbrock replied:
fixed

Fryguy commented:
How do you get this working?

I did the following locally
- `p = ProviderForeman.create!(:name => "local", :verify_ssl => false, :zone => Zone.default_zone, :url => hostname)`
  - Incidentally, the Provider should not have url alone.  It will need hostname, ipaddress, port, maybe protocol, and maybe path (if needed, but usually that can be hardcoded).
- `EmsRefresh.refresh(p.configuration_manager)` - when I do this I get ` undefined method 'my_zone' for #<ProvisioningManagerForeman:0x007f927ac724a0> ... ems_refresh.rb:113:in 'queue_merge'`, though it looks like the tables were all updated.  It looks like it's trying to put the provisioning manager refresh on the queue and failing.


cc @blomquisg @matthewd

brandondunne commented on lib/spec/vcr_cassettes/ManageiqForeman_Connection.yml:
```diff
@@ -0,0 +1,74 @@
+            "by": null,
+            "order": null
+          },
+          "results": [{"name":"abc.cloudforms.lab.eng.rdu2.redhat.com","id":29,"ip":"10.17.70.199","environment_id":1,"environment_name":"production","last_report":null,"mac":"78:2b:cb:00:f6:6c","realm_id":null,"realm_name":null,"sp_mac":"78:2b:cb:00:f6:6e","sp_ip":"10.17.70.210","sp_name":"","domain_id":1,"domain_name":"cloudforms.lab.eng.rdu2.redhat.com","architecture_id":1,"architecture_name": ...
```
@kbrock Please find / replace "cloudforms.lab.eng.rdu2.redhat.com" with "example.com".

brandondunne commented on lib/spec/vcr_cassettes/ManageiqForeman_Connection_all.yml:
```diff
@@ -0,0 +1,74 @@
+            "by": null,
+            "order": null
+          },
+          "results": [{"name":"abc.cloudforms.lab.eng.rdu2.redhat.com","id":29,"ip":"10.17.70.199","environment_id":1,"environment_name":"production","last_report":null,"mac":"78:2b:cb:00:f6:6c","realm_id":null,"realm_name":null,"sp_mac":"78:2b:cb:00:f6:6e","sp_ip":"10.17.70.210","sp_name":"","domain_id":1,"domain_name":"cloudforms.lab.eng.rdu2.redhat.com","architecture_id":1,"architecture_name": ...
```
Same here

brandondunne commented on vmdb/spec/models/ems_refresh/refreshers/foreman_refresher_spec.rb:
```diff
@@ -0,0 +1,131 @@
+    _guid, _server, zone = EvmSpecHelper.create_guid_miq_server_zone
+    FactoryGirl.create(:provider_foreman,
+                       :zone       => zone,
+                       :url        => "10.8.96.102",
```
This should be anonymized too.

brandondunne replied:
We need a better way to anonymize stuff like this

Fryguy replied:
We have a helper method for creating "random" IPs (see other factories or the factory_girl_helper).  However, if the test requires a specific IP for VCR, that should just be in the test like we do with all of the other refresher specs (and then anoynmized)

brandondunne commented:
@Fryguy "Incidentally, the Provider should not have url alone. It will need hostname, ipaddress, port, maybe protocol, and maybe path (if needed, but usually that can be hardcoded)."

Why would we need all of that?

blomquisg commented on vmdb/app/models/ems_refresh/save_inventory_helper.rb:
```diff
@@ -99,4 +99,35 @@ def store_ids_for_new_records(records, hashes, keys)
+  #   In Active Record, :dependent => :destory says the parent controls the lifespan of the children
+  # - We are targeting this association
+  #   If we are targeting something else, chances are it is a partial refresh. Don't delete.
+  #   If we are targeting this node, or targeting anything (nil), then refresh.
```
"refresh" should be "delete"?

chessbyte commented on vmdb/app/models/ems_refresh/save_inventory_helper.rb:
```diff
@@ -99,4 +99,35 @@ def store_ids_for_new_records(records, hashes, keys)
+  # This generates the "deletes" values based upon this intent
+  # It will delete missing records if both of the following is true:
+  # - The association is declared as a top_level association
+  #   In Active Record, :dependent => :destory says the parent controls the lifespan of the children
```
typo "destory"

jrafanie replied:
Looks like @kbrock borrowed my keyboard to type that one.  I do that one ALL THE TIME.

chessbyte commented on vmdb/app/models/ems_refresh/save_inventory_helper.rb:
```diff
@@ -99,4 +99,35 @@ def store_ids_for_new_records(records, hashes, keys)
+  # - make a complete refresh. Delete missing records.
+  # - make a partial refresh. Don't delete missing records.
+  # This generates the "deletes" values based upon this intent
+  # It will delete missing records if both of the following is true:
```
typo "both of the following IS true" should be "both of the following ARE true"

Fryguy commented:
> Why would we need all of that?

You don't for this task, but as a general "Provider", we will, especially as we start bucketing more things under provider.  Alternatively this would like in the multi-endpoint stuff.

When I tested locally, what value should go into the url attribute?  (answering this question affects UI too)  I can put the hostname without any https or anything and it works fine, so that makes me think it's really a hostname field an not a url field.

blomquisg commented:
So, I have two things.  Note that I don't wanna hold up merging this based on these two things.  But I wanna put them out there so that it's documented somewhere and maybe we can revisit later.

1)  I don't like the `ManageiqForeman::Connection` class.  The main thing is that there are several different ways to access different pieces of data.  Take `hostgroups` for example.  I can call:

* `Connection#hostgroups`
* `Connection#denormalized_hostgroups`
* `Connection#raw_hostgroups`

And, as a caller, I really don't know which is the right one to call when.  I'd like to see those methods simplified somehow.

But, here's the deal.  This is prototype for the first provider plugin.  So, I'm willing to say that what goes into your provider gem is your business.

2)  I really don't like [`missing_key`](https://github.com/ManageIQ/manageiq/pull/1388/files#diff-1b58bc3e51729f33f270712f9d11282cR5).  Specifically, I don't like the name.  I fully understand what purpose it's serving.  But, every time I read it, I have to remind myself what it's for.

For those that are reading this and don't know what it's for.  The `missing_key` is a flag set by the Foreman inventory parser and used by the Foreman Configuration Refresher to indicate that, while parsing the configuration inventory for foreman, the parser found a reference to something from the provisioning side that doesn't yet exist in our DB (i.e., we haven't collected that piece of provisioning inventory).  The Configuration Refresher uses this flag to know whether or not to queue up a Foreman Provisioning refresh.

That's a lot of English to describe that flag.  But, naming is hard, and I'm not being patronizing here.

Ultimately, I would really like to see these refreshers be able to handle targeted refreshes (i.e., refresh individual inventory objects).  Then, during the inventory parsing, it could simply queue up a refresh for the missing inventory object (by `manager_ref` value).

Again, I don't wanna pin this PR to these two points, but I want these points out in the wild, so I can re-reference them later.

Feel free to tear this up appropriately.

Fryguy commented:
How about `:needs_provisioning_refresh` for the key name

kbrock commented:
Ok,

connection:
- removed all raw methods / moved class lookup into hash
- dried up resource lookup
- added tests to exercise all api calls

others:
- remove pry from Gemfile
- removed duplicate vcr defs in gemfile
- renamed `foreman_spec` to `spec:foreman`
- fixed typos in save_inventory_helper
- renamed missing_key to needs_provisioning_refresh
- replaced ip addresses with example.com

chessbyte commented:
@kbrock Travis failed for some reason

blomquisg commented:
I rekicked the travis run just now ... maybe that will help.  It's failing on [Ems Refresh Status changes](https://github.com/ManageIQ/manageiq/pull/777).

kbrock commented:
Thanks @blomquisg / @chessbyte 

I had rebased this PR, which picked up the new ems refresh status columns code, and that failed.

Added a those columns to the managers.

Fryguy commented:
Why would those columns on a completely different class mess up this PR?

kbrock commented:
@Fryguy The refresh code sets those column values. The columns were added to `ExtManagementSystem`.
We refresh by `Manager`, and the code blew up when trying to set the columns that we did not have.
I added the column so people would know about refresh status of the Foreman `Managers` too.

Fryguy commented:
Oh I see.  For some reason I thought your refresher explicitly didn't inherit from BaseRefresher.  Maybe the code was that way on an initial iteration...oh well.  This is better anyway, and will be needed moving forward for the UI anyway.

chessbyte commented:
@kbrock @Fryguy what else is left to do/review on this PR?  Want to get this merged, if its good.

Fryguy commented:
Ok, so the queue_refresh is broken, but we will work on that in a separate PR... To refresh an EMS manually you would do:

```
p = ProviderForeman.create!(:name => "local", :verify_ssl => false, :zone => Zone.default_zone, :url => hostname)
EmsRefresh.refresh(p.configuration_manager)  # Will give an error but you will get the data
EmsRefresh.refresh(p.provisioning_manager)
```

jrafanie commented:
:tada: :star: :heart: Congratulations!!! This was a big PR.
